core/network-peerset/src/lib.rs

```diff
@@ -0,0 +1,689 @@
+			return false
+		}
+
+		let mut addrs: Vec<_> = addrs.collect();
+
+		if addrs.len() > 40 {
+			warn!(target: "sub-libp2p", "Attempt to add more than 40 addresses for {:?}", peer_id);
+			addrs.truncate(40);
+		}
+
+		let now_systime = SystemTime::now();
+		let now = Instant::now();
+
+		let peer = peer_access(&mut self.store, peer_id);
+
+		let new_addrs = peer.addrs
+			.drain(..)
+			.filter(|a| {
+				if a.expires < now_systime && !a.is_connected() {
+					return false
+				}
+				addrs.retain(|(addr, _)| *addr != a.addr);
+				true
+			})
+			.collect();
+		peer.addrs = new_addrs;
+
+		let mut anything_changed = false;
+
+		if !addrs.is_empty() {
+			anything_changed = true;
+			trace!(
+				target: "sub-libp2p",
+				"Peer store: adding addresses {:?} for {:?}",
+				addrs,
+				peer_id,
+			);
+		}
+
+		'addrs_inserter: for (addr, connectable) in addrs {
+			let initial_score = if connectable {
+				DISCOVERY_INITIAL_SCORE_CONNECTABLE
+			} else {
+				DISCOVERY_INITIAL_SCORE
+			};
+
+			// Enforce `MAX_ADDRESSES_PER_PEER` before inserting, or skip this entry.
+			while peer.addrs.len() >= MAX_ADDRESSES_PER_PEER {
+				let pos = peer.addrs.iter_mut().position(|addr| addr.score() <= initial_score);
+				if let Some(pos) = pos {
+					let _ = peer.addrs.remove(pos);
+				} else {
+					continue 'addrs_inserter;
+				}
+			}
+
+			// `addrs` can contain duplicates, therefore we would insert the same address twice.
+			if peer.addrs.iter().any(|a| a.addr == addr) {
+				continue;
+			}
+
+			peer.addrs.push(Addr {
+				addr,
+				expires: now_systime + KADEMLIA_DISCOVERY_EXPIRATION,
+				back_off_until: now,
+				next_back_off: FIRST_CONNECT_FAIL_BACKOFF,
+				score: AddrScore {
+					connected_since: None,
+					score: initial_score,
+					latest_score_update: now,
+				},
+			});
+		}
+
+		anything_changed
+	}
+
+	/// Modifies the score of the given peer.
+	pub fn report_peer(&mut self, _peer_id: &PeerId, _score_modif: i32) {
+		// TODO: not implemented
+	}
+
+	/// Returns the list of peers that are stored in the topology.
+	pub fn known_peers(&self) -> impl Iterator<Item = &PeerId> {
+		self.store.keys()
+	}
+
+	/// Returns the addresses stored for a specific peer, and their reputation score.
+	///
+	/// If `include_expired` is true, includes expired addresses that shouldn't be taken into
+	/// account when dialing.
+	pub fn addresses_of_peer(&mut self, peer: &PeerId, include_expired: bool)
+		-> impl Iterator<Item = (&Multiaddr, u32)> {
+		let now_st = SystemTime::now();
+		let now_is = Instant::now();
+
+		let mut list = self.store.get_mut(peer).into_iter().flat_map(|p| p.addrs.iter_mut()).filter_map(move |addr| {
+			let (score, connected) = addr.score_and_is_connected();
+			if include_expired || (addr.expires >= now_st && score > 0 && addr.back_off_until < now_is) || connected {
+				Some((score, &addr.addr))
+			} else {
+				None
+			}
+		}).collect::<Vec<_>>();
+		list.sort_by(|a, b| a.0.cmp(&b.0));
+		// TODO: meh, optimize
```

Comment:
```suggestion
		// This isn't optimal and should probably be optimised.
```

No `TODO` without link/ticket ;)